File: api/sdks/ruby/lib/objstore/clients/quic_client.rb
require "net/http"
require "json"
require "openssl"

module ObjectStore
  module Clients
    # QUIC/HTTP3 client implementation for ObjectStore
    # Note: Ruby doesn't have native HTTP/3 support yet, so this uses HTTP/2 or HTTP/1.1
    # with Net::HTTP as a fallback implementation
    #
    # @api private
    class QuicClient
      attr_reader :host, :port, :use_ssl, :timeout

      # Initialize a new QUIC client
      #
      # @param host [String] Server hostname
      # @param port [Integer] Server port (default: 4433 for QUIC)
      # @param use_ssl [Boolean] Whether to use TLS
      # @param timeout [Integer] Request timeout in seconds
      # @param token [String, nil] Bearer token for Authorization header
      # @param headers [Hash] Additional HTTP headers to send on every request
      # @param tenant_id [String, nil] Tenant identifier sent as X-Tenant-ID header
      def initialize(host: "localhost", port: 4433, use_ssl: true, timeout: 30,
                     token: nil, headers: {}, tenant_id: nil, verify_ssl: true)
        @host = host
        @port = port
        @use_ssl = use_ssl
        @timeout = timeout
        @token = token
        @extra_headers = headers || {}
        @tenant_id = tenant_id
        # TLS certificates are verified by default; disable only for testing
        # against self-signed certificates.
        @verify_mode = verify_ssl ? OpenSSL::SSL::VERIFY_PEER : OpenSSL::SSL::VERIFY_NONE
      end

      # Upload an object to the store
      #
      # @param key [String] Object key
      # @param data [String] Object data
      # @param metadata [Models::Metadata, Hash, nil] Optional metadata
      #
      # @return [Models::PutResponse] Response with success status and etag
      #
      # @raise [ValidationError, ServerError]
      def put(key, data, metadata = nil)
        metadata_obj = metadata.is_a?(Models::Metadata) ? metadata : Models::Metadata.new(metadata || {})

        uri = build_uri("/objects/#{encode_key(key)}")
        request = Net::HTTP::Put.new(uri)
        request.body = data
        request["Content-Type"] = metadata_obj.content_type || "application/octet-stream"

        if metadata_obj.content_encoding
          request["Content-Encoding"] = metadata_obj.content_encoding
        end

        # Custom metadata is sent as individual X-Meta-<key> headers (the QUIC
        # server reads Content-Type/Content-Encoding plus X-Meta-* headers).
        if metadata_obj.custom
          metadata_obj.custom.each do |k, v|
            request["X-Meta-#{k}"] = v.to_s
          end
        end

        response = execute_request(uri, request)

        Models::PutResponse.new(
          success: true,
          message: parse_body(response)["message"],
          etag: response["etag"]
        )
      end

      # Upload an object from an IO stream
      #
      # @param key [String] Object key
      # @param io [IO, #read] IO object to read from
      # @param metadata [Models::Metadata, Hash, nil] Optional metadata
      # @param chunk_size [Integer] Size of chunks to read
      #
      # @return [Models::PutResponse] Response with success status and etag
      #
      # @raise [ValidationError, ServerError]
      def put_stream(key, io, metadata: nil, chunk_size: 8192)
        metadata_obj = metadata.is_a?(Models::Metadata) ? metadata : Models::Metadata.new(metadata || {})

        uri = build_uri("/objects/#{encode_key(key)}")
        request = Net::HTTP::Put.new(uri)
        request["Content-Type"] = metadata_obj.content_type || "application/octet-stream"
        request["Transfer-Encoding"] = "chunked"

        if metadata_obj.content_encoding
          request["Content-Encoding"] = metadata_obj.content_encoding
        end

        # Custom metadata is sent as individual X-Meta-<key> headers.
        if metadata_obj.custom
          metadata_obj.custom.each do |k, v|
            request["X-Meta-#{k}"] = v.to_s
          end
        end

        # Set request body to the IO stream
        request.body_stream = io

        response = execute_request(uri, request)

        Models::PutResponse.new(
          success: true,
          message: parse_body(response)["message"],
          etag: response["etag"]
        )
      end

      # Retrieve an object from the store
      #
      # @param key [String] Object key
      #
      # @return [Models::GetResponse] Response with data and metadata
      #
      # @raise [NotFoundError, ServerError]
      def get(key)
        uri = build_uri("/objects/#{encode_key(key)}")
        request = Net::HTTP::Get.new(uri)

        response = execute_request(uri, request)

        metadata = metadata_from_headers(response, size: response.body&.bytesize)

        Models::GetResponse.new(response.body, metadata)
      end

      # Retrieve an object in chunks via streaming
      #
      # @param key [String] Object key
      #
      # @yieldparam chunk [String] A chunk of data
      #
      # @return [Models::Metadata] Object metadata
      #
      # @raise [NotFoundError, ServerError]
      def get_stream(key, chunk_size: 8192)
        uri = build_uri("/objects/#{encode_key(key)}")
        request = Net::HTTP::Get.new(uri)

        metadata = nil
        apply_auth_headers(request)

        Net::HTTP.start(uri.hostname, uri.port,
                        use_ssl: @use_ssl,
                        read_timeout: @timeout,
                        write_timeout: @timeout,
                        verify_mode: @verify_mode) do |http|
          http.request(request) do |response|
            handle_response(response)

            # Extract metadata from headers
            metadata = Models::Metadata.new(
              content_type: response["content-type"],
              size: response["content-length"]&.to_i,
              etag: response["etag"],
              last_modified: response["last-modified"]
            )

            # Stream the response body in chunks
            response.read_body do |chunk|
              yield chunk if block_given?
            end
          end
        end

        metadata
      rescue Net::OpenTimeout, Net::ReadTimeout => e
        raise ObjectStore::TimeoutError, e.message
      rescue ObjectStore::Error
        raise
      rescue StandardError => e
        raise ObjectStore::ConnectionError, e.message
      end

      # Delete an object from the store
      #
      # @param key [String] Object key
      #
      # @return [Models::DeleteResponse] Response with success status
      #
      # @raise [NotFoundError, ServerError]
      def delete(key)
        uri = build_uri("/objects/#{encode_key(key)}")
        request = Net::HTTP::Delete.new(uri)

        response = execute_request(uri, request)

        # The server returns 204 No Content; parse_body yields {} for the
        # empty body, so default the message.
        body = parse_body(response)
        message = body.is_a?(Hash) ? body["message"] : nil
        Models::DeleteResponse.new(success: true, message: message || "Object deleted successfully")
      end

      def list(prefix: nil, delimiter: nil, max_results: 100, continue_from: nil)
        params = {}
        params[:prefix] = prefix if prefix
        params[:delimiter] = delimiter if delimiter
        params[:max] = max_results if max_results
        params[:continue] = continue_from if continue_from

        uri = build_uri("/objects", params)
        request = Net::HTTP::Get.new(uri)

        response = execute_request(uri, request)
        body = parse_body(response)

        Models::ListResponse.new(body)
      end

      def exists?(key)
        uri = build_uri("/objects/#{encode_key(key)}")
        request = Net::HTTP::Head.new(uri)

        response = execute_request(uri, request)
        response.code == "200"
      rescue ObjectStore::NotFoundError
        # If we get a 404 for a HEAD request, object doesn't exist
        false
      end

      def get_metadata(key)
        # The QUIC server has no /metadata route; metadata is read via a
        # HEAD on the object and parsed from response headers.
        uri = build_uri("/objects/#{encode_key(key)}")
        request = Net::HTTP::Head.new(uri)

        response = execute_request(uri, request)

        Models::MetadataResponse.new(
          metadata: metadata_from_headers(response),
          success: true
        )
      end

      def update_metadata(key, metadata)
        metadata_obj = metadata.is_a?(Models::Metadata) ? metadata : Models::Metadata.new(metadata)

        # The QUIC server updates metadata via PATCH on the object with a
        # JSON body of { content_type, content_encoding, custom }.
        payload = {
          content_type: metadata_obj.content_type,
          content_encoding: metadata_obj.content_encoding,
          custom: metadata_obj.custom
        }.compact

        uri = build_uri("/objects/#{encode_key(key)}")
        request = Net::HTTP::Patch.new(uri)
        request["Content-Type"] = "application/json"
        request.body = payload.to_json

        response = execute_request(uri, request)

        Models::UpdateMetadataResponse.new(success: true, message: parse_body(response)["message"])
      end

      def health(service: nil)
        params = service ? { service: service } : {}
        uri = build_uri("/health", params)
        request = Net::HTTP::Get.new(uri)

        response = execute_request(uri, request)
        body = parse_body(response)

        Models::HealthResponse.new(body)
      end

      def archive(key, destination_type:, destination_settings: {})
        payload = {
          key: key,
          destination_type: destination_type,
          destination_settings: destination_settings
        }

        uri = build_uri("/archive")
        request = Net::HTTP::Post.new(uri)
        request["Content-Type"] = "application/json"
        request.body = payload.to_json

        response = execute_request(uri, request)

        Models::ArchiveResponse.new(success: true, message: parse_body(response)["message"])
      end

      def add_policy(policy)
        policy_obj = policy.is_a?(Models::LifecyclePolicy) ? policy : Models::LifecyclePolicy.new(policy)

        uri = build_uri("/policies")
        request = Net::HTTP::Post.new(uri)
        request["Content-Type"] = "application/json"
        request.body = policy_obj.to_json

        response = execute_request(uri, request)

        { success: true, message: parse_body(response)["message"] }
      end

      def remove_policy(id)
        uri = build_uri("/policies/#{id}")
        request = Net::HTTP::Delete.new(uri)

        response = execute_request(uri, request)

        { success: true, message: parse_body(response)["message"] }
      end

      def get_policies(prefix: nil)
        params = prefix ? { prefix: prefix } : {}
        uri = build_uri("/policies", params)
        request = Net::HTTP::Get.new(uri)

        response = execute_request(uri, request)
        body = parse_body(response)

        policies = (body["policies"] || []).map { |p| Models::LifecyclePolicy.new(p) }
        { policies: policies, success: true }
      end

      def apply_policies
        uri = build_uri("/policies/apply")
        request = Net::HTTP::Post.new(uri)

        response = execute_request(uri, request)
        body = parse_body(response)

        {
          success: true,
          policies_count: body["policies_count"] || 0,
          objects_processed: body["objects_processed"] || 0,
          message: body["message"]
        }
      end

      def add_replication_policy(policy)
        policy_obj = policy.is_a?(Models::ReplicationPolicy) ? policy : Models::ReplicationPolicy.new(policy)

        # The QUIC server expects `check_interval` (seconds), not
        # `check_interval_seconds` as the REST server uses.
        payload = policy_obj.to_h
        if payload.key?(:check_interval_seconds)
          payload[:check_interval] = payload.delete(:check_interval_seconds)
        end

        uri = build_uri("/replication/policies")
        request = Net::HTTP::Post.new(uri)
        request["Content-Type"] = "application/json"
        request.body = payload.to_json

        response = execute_request(uri, request)

        { success: true, message: parse_body(response)["message"] }
      end

      def remove_replication_policy(id)
        uri = build_uri("/replication/policies/#{id}")
        request = Net::HTTP::Delete.new(uri)

        response = execute_request(uri, request)

        { success: true, message: parse_body(response)["message"] }
      end

      def get_replication_policies
        uri = build_uri("/replication/policies")
        request = Net::HTTP::Get.new(uri)

        response = execute_request(uri, request)
        body = parse_body(response)

        policies = (body["policies"] || []).map { |p| Models::ReplicationPolicy.new(p) }
        { policies: policies }
      end

      def get_replication_policy(id)
        uri = build_uri("/replication/policies/#{id}")
        request = Net::HTTP::Get.new(uri)

        response = execute_request(uri, request)
        body = parse_body(response)

        { policy: Models::ReplicationPolicy.new(body) }
      end

      def trigger_replication(policy_id: nil, parallel: false, worker_count: 4)
        # The QUIC server takes policy_id as a query param (empty = sync all),
        # not a JSON body. parallel/worker_count are not supported over QUIC.
        params = {}
        params[:policy_id] = policy_id if policy_id

        uri = build_uri("/replication/trigger", params)
        request = Net::HTTP::Post.new(uri)

        response = execute_request(uri, request)
        body = parse_body(response)

        {
          success: true,
          result: body["result"]&.transform_keys(&:to_sym),
          message: body["message"]
        }
      end

      def get_replication_status(id)
        uri = build_uri("/replication/status/#{id}")
        request = Net::HTTP::Get.new(uri)

        response = execute_request(uri, request)
        body = parse_body(response)

        # The server returns status fields flat at the top level (not nested
        # under a "status" key).
        {
          success: true,
          status: Models::ReplicationStatus.new(body)
        }
      end

      # Close the underlying HTTP connection if one is held
      #
      # This client opens a short-lived connection per request via
      # Net::HTTP.start, so there is no persistent connection to release.
      # Provided for API parity; closes a persistent connection if one is
      # ever held. Safe to call multiple times.
      #
      # @return [void]
      def close
        return unless defined?(@http) && @http

        @http.finish if @http.respond_to?(:finish) && @http.respond_to?(:started?) && @http.started?
        @http = nil
      end

      private

      def build_uri(path, params = {})
        scheme = @use_ssl ? "https" : "http"
        uri = URI("#{scheme}://#{@host}:#{@port}#{path}")

        unless params.empty?
          uri.query = URI.encode_www_form(params)
        end

        uri
      end

      def encode_key(key)
        URI.encode_www_form_component(key)
      end

      # Build a Models::Metadata from the standard and X-Meta-<key> response
      # headers. The QUIC server conveys content_type/content_encoding via
      # standard headers and custom metadata as individual X-Meta-* headers.
      #
      # @param response [Net::HTTPResponse] the HTTP response to read headers from
      # @param size [Integer, nil] explicit size override (e.g. body bytesize for
      #   GET, where Content-Length may be absent); falls back to Content-Length
      #
      # @return [Models::Metadata]
      def metadata_from_headers(response, size: nil)
        custom = {}
        response.each_header do |name, value|
          next unless name =~ /\Ax-meta-(.+)\z/i

          custom[Regexp.last_match(1)] = value
        end

        Models::Metadata.new(
          content_type: response["content-type"],
          content_encoding: response["content-encoding"],
          size: size || response["content-length"]&.to_i,
          etag: response["etag"],
          last_modified: response["last-modified"],
          custom: custom
        )
      end

      def apply_auth_headers(request)
        request["Authorization"] = "Bearer #{@token}" if @token
        request["X-Tenant-ID"] = @tenant_id if @tenant_id
        @extra_headers.each { |k, v| request[k.to_s] = v.to_s }
      end

      def execute_request(uri, request)
        apply_auth_headers(request)
        Net::HTTP.start(uri.hostname, uri.port,
                        use_ssl: @use_ssl,
                        read_timeout: @timeout,
                        write_timeout: @timeout,
                        verify_mode: @verify_mode) do |http|
          response = http.request(request)
          handle_response(response)
        end
      rescue Net::OpenTimeout, Net::ReadTimeout => e
        raise ObjectStore::TimeoutError, e.message
      rescue ObjectStore::Error
        # Re-raise our custom errors as-is
        raise
      rescue StandardError => e
        raise ObjectStore::ConnectionError, e.message
      end

      def handle_response(response)
        case response.code
        when "200", "201", "204"
          response
        when "404"
          raise ObjectStore::NotFoundError, "Resource not found"
        when "400"
          raise ObjectStore::ValidationError, error_message(response, "Bad request")
        when "401"
          raise ObjectStore::AuthenticationError, error_message(response, "Unauthenticated")
        when "403"
          raise ObjectStore::AuthorizationError, error_message(response, "Forbidden")
        when "409"
          raise ObjectStore::AlreadyExistsError, error_message(response, "Already exists")
        when "413"
          raise ObjectStore::ValidationError, "Request entity too large"
        when "429"
          raise ObjectStore::RateLimitError, error_message(response, "Rate limited")
        when /^5/
          raise ObjectStore::ServerError, error_message(response, "Server error")
        else
          raise ObjectStore::Error, "Unexpected response: #{response.code}"
        end
      end

      # Extract the server-provided error message from a response body,
      # falling back to a generic default when the body is absent or not JSON.
      def error_message(response, fallback)
        body = parse_body(response)
        (body.is_a?(Hash) && body["message"]) || fallback
      end

      def parse_body(response)
        return {} if response.body.nil? || response.body.empty?
        return response.body unless response["content-type"]&.include?("application/json")

        JSON.parse(response.body)
      rescue JSON::ParserError
        {}
      end
    end
  end
end

File: api/sdks/javascript/src/clients/QuicClient.js
// Check for Node.js environment
const isNode = typeof process !== 'undefined' && process.versions && process.versions.node;

/**
 * Convert replication mode to string format for REST API.
 * @param {number|string} mode - Replication mode (0, 1, 'transparent', 'opaque')
 * @returns {string} String representation ('transparent' or 'opaque')
 */
function normalizeReplicationMode(mode) {
  if (typeof mode === 'number') {
    return mode === 1 ? 'opaque' : 'transparent';
  }
  if (typeof mode === 'string') {
    return mode.toLowerCase() === 'opaque' ? 'opaque' : 'transparent';
  }
  return 'transparent';
}

/**
 * QUIC/HTTP3 client for go-objstore
 * Supports all object storage operations via QUIC/HTTP3 protocol
 *
 * Note: Full HTTP/3 support in Node.js is still experimental.
 * This implementation uses HTTP/2 over TLS as a fallback with the same API surface.
 * In browser environments, uses standard fetch API.
 */
export class QuicClient {
  /**
   * Create a new QUIC/HTTP3 client
   * @param {Object} config - Client configuration
   * @param {string} config.baseURL - Base URL of the object store server
   * @param {number} [config.timeout=30000] - Request timeout in milliseconds
   * @param {Object} [config.headers={}] - Additional headers for all requests
   */
  constructor(config) {
    if (!config?.baseURL) {
      throw new Error('baseURL is required');
    }

    this.baseURL = config.baseURL.replace(/\/$/, '');
    this.timeout = config.timeout || 30000;
    this.headers = config.headers || {};
    this.isNode = isNode;

    // Note: HTTPS agent is not created here for better browser compatibility
    // The fetch API will handle HTTPS connections automatically
    this.agent = null;
  }

  /**
   * Make an HTTP request using fetch API
   * @private
   */
  async _makeRequest(method, path, options = {}) {
    const url = `${this.baseURL}${path}`;

    const headers = {
      ...this.headers,
      ...options.headers,
    };

    const fetchOptions = {
      method,
      headers,
      signal: AbortSignal.timeout(this.timeout),
    };

    // Note: Agent is not used for better browser compatibility
    // Modern fetch API handles HTTPS automatically

    if (options.body !== undefined) {
      fetchOptions.body = options.body;
    }

    const response = await fetch(url, fetchOptions);

    if (!response.ok && response.status !== 404) {
      const errorText = await response.text();
      throw new Error(`QUIC/HTTP3 error (${response.status}): ${errorText}`);
    }

    return response;
  }

  /**
   * Store an object in the backend
   * @param {string} key - Storage key for the object
   * @param {Buffer|Uint8Array|string} data - Object data
   * @param {Object} [metadata] - Optional metadata
   * @returns {Promise<Object>} Response with success status and etag
   */
  async put(key, data, metadata = null) {
    if (!key) throw new Error('key is required');
    if (!data) throw new Error('data is required');

    const headers = {
      'Content-Type': metadata?.contentType || 'application/octet-stream',
    };

    if (metadata?.contentEncoding) {
      headers['Content-Encoding'] = metadata.contentEncoding;
    }

    if (metadata?.custom) {
      headers['X-Metadata'] = JSON.stringify(metadata.custom);
    }

    const response = await this._makeRequest(
      'PUT',
      `/objects/${encodeURIComponent(key)}`,
      {
        headers,
        body: Buffer.from(data),
      }
    );

    const result = await response.json();
    return {
      success: true,
      message: result.message || 'Object uploaded successfully',
      etag: result.data?.etag || result.etag || '',
    };
  }

  /**
   * Retrieve an object from the backend
   * @param {string} key - Storage key for the object
   * @returns {Promise<Object>} Response with data and metadata
   */
  async get(key) {
    if (!key) throw new Error('key is required');

    const response = await this._makeRequest('GET', `/objects/${encodeURIComponent(key)}`);

    const metadata = {
      contentType: response.headers.get('content-type'),
      contentLength: parseInt(response.headers.get('content-length') || '0'),
      etag: response.headers.get('etag'),
      lastModified: response.headers.get('last-modified'),
    };

    const arrayBuffer = await response.arrayBuffer();
    return {
      data: Buffer.from(arrayBuffer),
      metadata,
    };
  }

  /**
   * Delete an object from the backend
   * @param {string} key - Storage key for the object
   * @returns {Promise<Object>} Response with success status
   */
  async delete(key) {
    if (!key) throw new Error('key is required');

    const response = await this._makeRequest('DELETE', `/objects/${encodeURIComponent(key)}`);
    const result = await response.json();

    return {
      success: true,
      message: result.message || 'Object deleted successfully',
    };
  }

  /**
   * List objects matching the criteria
   * @param {Object} [options] - List options
   * @param {string} [options.prefix] - Prefix to filter objects
   * @param {string} [options.delimiter] - Delimiter for hierarchical listing
   * @param {number} [options.limit] - Maximum number of results
   * @param {string} [options.token] - Pagination token
   * @returns {Promise<Object>} Response with objects list
   */
  async list(options = {}) {
    const params = new URLSearchParams();
    if (options.prefix) params.append('prefix', options.prefix);
    if (options.delimiter) params.append('delimiter', options.delimiter);
    if (options.limit) params.append('limit', options.limit.toString());
    if (options.token) params.append('token', options.token);

    const queryString = params.toString();
    const path = queryString ? `/objects?${queryString}` : '/objects';
    const response = await this._makeRequest('GET', path);
    const result = await response.json();

    return {
      objects: result.objects || [],
      commonPrefixes: result.common_prefixes || [],
      nextToken: result.next_token || '',
      truncated: result.truncated || false,
    };
  }

  /**
   * List objects with advanced options
   * Provides explicit control over all listing parameters
   * @param {Object} [options] - List options
   * @param {string} [options.prefix=''] - Prefix to filter objects
   * @param {string} [options.delimiter=''] - Delimiter for hierarchical listing
   * @param {number} [options.maxResults=0] - Maximum number of results (0 = unlimited)
   * @param {string} [options.continueFrom=''] - Continuation token for pagination
   * @returns {Promise<Object>} Response with objects, commonPrefixes, nextToken, and truncated flag
   */
  async listWithOptions(options = {}) {
    // Validate options
    if (options.prefix !== undefined && typeof options.prefix !== 'string') {
      throw new Error('prefix must be a string');
    }
    if (options.delimiter !== undefined && typeof options.delimiter !== 'string') {
      throw new Error('delimiter must be a string');
    }
    if (options.maxResults !== undefined && typeof options.maxResults !== 'number') {
      throw new Error('maxResults must be a number');
    }
    if (options.continueFrom !== undefined && typeof options.continueFrom !== 'string') {
      throw new Error('continueFrom must be a string');
    }

    const params = new URLSearchParams();
    if (options.prefix) params.append('prefix', options.prefix);
    if (options.delimiter) params.append('delimiter', options.delimiter);
    if (options.maxResults) params.append('limit', options.maxResults.toString());
    if (options.continueFrom) params.append('token', options.continueFrom);

    const queryString = params.toString();
    const path = queryString ? `/objects?${queryString}` : '/objects';
    const response = await this._makeRequest('GET', path);
    const result = await response.json();

    return {
      objects: result.objects || [],
      commonPrefixes: result.common_prefixes || [],
      nextToken: result.next_token || '',
      truncated: result.truncated || false,
    };
  }

  /**
   * Check if an object exists
   * @param {string} key - Storage key for the object
   * @returns {Promise<Object>} Response with exists status
   */
  async exists(key) {
    if (!key) throw new Error('key is required');

    try {
      const response = await this._makeRequest('HEAD', `/objects/${encodeURIComponent(key)}`);
      return { exists: response.ok };
    } catch (error) {
      return { exists: false };
    }
  }

  /**
   * Get metadata for an object
   * @param {string} key - Storage key for the object
   * @returns {Promise<Object>} Response with metadata
   */
  async getMetadata(key) {
    if (!key) throw new Error('key is required');

    const response = await this._makeRequest('GET', `/metadata/${encodeURIComponent(key)}`);
    const result = await response.json();

    return {
      success: true,
      metadata: result.metadata || {},
    };
  }

  /**
   * Update metadata for an object
   * @param {string} key - Storage key for the object
   * @param {Object} metadata - New metadata
   * @returns {Promise<Object>} Response with success status
   */
  async updateMetadata(key, metadata) {
    if (!key) throw new Error('key is required');
    if (!metadata) throw new Error('metadata is required');

    const response = await this._makeRequest(
      'PUT',
      `/metadata/${encodeURIComponent(key)}`,
      {
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(metadata),
      }
    );

    const result = await response.json();
    return {
      success: true,
      message: result.message || 'Metadata updated successfully',
    };
  }

  /**
   * Health check
   * @returns {Promise<Object>} Health status
   */
  async health() {
    const response = await this._makeRequest('GET', '/health');
    const result = await response.json();

    return {
      status: result.status || 'unknown',
      message: result.message || '',
    };
  }

  /**
   * Archive an object to archival storage
   * @param {string} key - Storage key for the object
   * @param {string} destinationType - Destination backend type
   * @param {Object} [destinationSettings] - Destination backend settings
   * @returns {Promise<Object>} Response with success status
   */
  async archive(key, destinationType, destinationSettings = {}) {
    if (!key) throw new Error('key is required');
    if (!destinationType) throw new Error('destinationType is required');

    const response = await this._makeRequest('POST', '/archive', {
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        key,
        destination_type: destinationType,
        destination_settings: destinationSettings,
      }),
    });

    const result = await response.json();
    return {
      success: true,
      message: result.message || 'Object archived successfully',
    };
  }

  /**
   * Add a lifecycle policy
   * @param {Object} policy - Lifecycle policy
   * @returns {Promise<Object>} Response with success status
   */
  async addPolicy(policy) {
    if (!policy) throw new Error('policy is required');

    const response = await this._makeRequest('POST', '/policies', {
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(policy),
    });

    const result = await response.json();
    return {
      success: true,
      message: result.message || 'Policy added successfully',
    };
  }

  /**
   * Remove a lifecycle policy
   * @param {string} id - Policy ID
   * @returns {Promise<Object>} Response with success status
   */
  async removePolicy(id) {
    if (!id) throw new Error('id is required');

    const response = await this._makeRequest('DELETE', `/policies/${id}`);
    const result = await response.json();

    return {
      success: true,
      message: result.message || 'Policy removed successfully',
    };
  }

  /**
   * Get all lifecycle policies
   * @param {string} [prefix] - Optional prefix filter
   * @returns {Promise<Object>} Response with policies list
   */
  async getPolicies(prefix = '') {
    const path = prefix ? `/policies?prefix=${encodeURIComponent(prefix)}` : '/policies';
    const response = await this._makeRequest('GET', path);
    const result = await response.json();

    return {
      success: true,
      policies: result.policies || [],
    };
  }

  /**
   * Apply all lifecycle policies
   * @returns {Promise<Object>} Response with apply results
   */
  async applyPolicies() {
    const response = await this._makeRequest('POST', '/policies/apply', {
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({}),
    });

    const result = await response.json();
    return {
      success: true,
      policiesCount: result.policies_count || 0,
      objectsProcessed: result.objects_processed || 0,
      message: result.message || '',
    };
  }

  /**
   * Add a replication policy
   * @param {Object} policy - Replication policy
   * @returns {Promise<Object>} Response with success status
   */
  async addReplicationPolicy(policy) {
    if (!policy) throw new Error('policy is required');

    // Normalize replication_mode to string format expected by REST API
    const normalizedPolicy = {
      ...policy,
      replication_mode: normalizeReplicationMode(policy.replication_mode || policy.replicationMode),
    };

    const response = await this._makeRequest('POST', '/replication/policies', {
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(normalizedPolicy),
    });

    const result = await response.json();
    return {
      success: true,
      message: result.message || 'Replication policy added successfully',
    };
  }

  /**
   * Remove a replication policy
   * @param {string} id - Policy ID
   * @returns {Promise<Object>} Response with success status
   */
  async removeReplicationPolicy(id) {
    if (!id) throw new Error('id is required');

    const response = await this._makeRequest('DELETE', `/replication/policies/${id}`);
    const result = await response.json();

    return {
      success: true,
      message: result.message || 'Replication policy removed successfully',
    };
  }

  /**
   * Get all replication policies
   * @returns {Promise<Object>} Response with policies list
   */
  async getReplicationPolicies() {
    const response = await this._makeRequest('GET', '/replication/policies');
    const result = await response.json();

    return {
      policies: result.policies || [],
    };
  }

  /**
   * Get a specific replication policy
   * @param {string} id - Policy ID
   * @returns {Promise<Object>} Response with policy
   */
  async getReplicationPolicy(id) {
    if (!id) throw new Error('id is required');

    const response = await this._makeRequest('GET', `/replication/policies/${id}`);
    const result = await response.json();

    return {
      policy: result.policy || null,
    };
  }

  /**
   * Trigger replication sync
   * @param {Object} [options] - Trigger options
   * @param {string} [options.policyId] - Policy ID to sync (empty for all)
   * @param {boolean} [options.parallel] - Use parallel workers
   * @param {number} [options.workerCount] - Number of workers
   * @returns {Promise<Object>} Response with sync results
   */
  async triggerReplication(options = {}) {
    const response = await this._makeRequest('POST', '/replication/trigger', {
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        policy_id: options.policyId || '',
        parallel: options.parallel || false,
        worker_count: options.workerCount || 4,
      }),
    });

    const result = await response.json();
    return {
      success: true,
      result: result.result || {},
      message: result.message || '',
    };
  }

  /**
   * Get replication status
   * @param {string} id - Policy ID
   * @returns {Promise<Object>} Response with replication status
   */
  async getReplicationStatus(id) {
    if (!id) throw new Error('id is required');

    const response = await this._makeRequest('GET', `/replication/status/${id}`);
    const result = await response.json();

    return {
      success: true,
      status: result.status || {},
      message: result.message || '',
    };
  }

  /**
   * Close the client
   */
  close() {
    // No cleanup needed for fetch
  }
}
